### ISSUE TITLE
[Serve] Batched sparse decode double-counts the batch offset into the QSA block-mask row index

### ISSUE THREAD

**Issue by flaviusburca** [invergent-ai/surogate#110]
**Severity:** high &nbsp;·&nbsp; **Location:** [`csrc/src/serve/ops/kernel/gqa_attention_decode_bf16.cuh:378`](https://github.com/invergent-ai/surogate/blob/7f12387624e04c498dbd8df879eaeff13dea0b25/csrc/src/serve/ops/kernel/gqa_attention_decode_bf16.cuh#L378) &nbsp;·&nbsp; **Audited at:** `7f123876`

### Defect
In gqa_attention_small_t_tc_partial_bf16_kernel the QSA block-mask row is computed as `mask_base = batch*full_width + column_base`, but `column_base` already had `batch*full_width` folded into it 238 lines earlier (line 140). The batch offset is therefore applied twice, so the mask row used for batch b is `2*b*full_width + column_begin + token` instead of `b*full_width + column_begin + token`. The mask tensor is allocated in TextContext::text_indexer_selection as `work_.alloc(DType::I32, {words, tokens})` with `tokens = width*batch` and `stride = words`, i.e. exactly one row per global query column, so any b > 0 either reads another sequence's selection or reads past the end of the allocation. Every other consumer of `column_base` (q, pos, input.k/v, partial_acc/m/l) uses it directly and is correct; only the mask adds the batch term a second time. The prefill twin (gqa_attention_prefill_bf16.cuh, `block_mask.words + qrow0 * block_mask.stride`) is correct because its launcher advances `mask.words` per batch slice.

### Failure scenario
A QSA-indexer target (V::has_qsa_indexer) running a 4-sequence decode round of one token each past the indexer budget: ops::gqa_attention is called with q of shape [D,Hq,1,4] and a non-null selection, gqa_attention_resolve_route returns SmallT, and gqa_attention_small_t_launch_for instantiates the kernel with MultiBatch=true, Sparse=true. full_width=1, column_begin=0. Batch 0 reads mask row 0 (correct); batch 1 reads mask row 2 — sequence 2's block selection, so sequence 1 attends to the wrong set of KV blocks and produces wrong logits; batch 2 reads row 4 and batch 3 reads row 6, both past the end of the 4-row mask tensor, reading whatever the workspace arena holds after it. The result is a wrong-token generation for lanes 1-3 plus an out-of-bounds global read that can fault when the mask sits at the tail of a workspace allocation.

### Evidence
````cpp
gqa_attention_decode_bf16.cuh:139-140
    std::int64_t column_base = column_begin;
    if constexpr (MultiBatch) { column_base += static_cast<std::int64_t>(batch) * full_width; }

gqa_attention_decode_bf16.cuh:377-386
        const std::int64_t mask_base =
            static_cast<std::int64_t>(MultiBatch ? batch : 0) * full_width + column_base;
        const std::uint32_t* mask0 =
            Sparse && row0 < row_count
                ? block_mask.words + (mask_base + token0) * block_mask.stride
                : nullptr;
        const std::uint32_t* mask1 =
            Sparse && row1 < row_count
                ? block_mask.words + (mask_base + token1) * block_mask.stride
                : nullptr;
````

### Suggested fix
Drop the redundant batch term: `const std::int64_t mask_base = column_base;` (column_base already equals column_begin + batch*full_width, which is exactly the [words, full_width, batch] row index the comment describes). Add a batched-sparse case to the serve tests — the existing QSA coverage appears to be single-sequence only, which is why this survived.

### Verification
Found in an audit of `csrc/src/serve`, then upheld by two independent verifiers: one tried to refute it, the other traced a concrete path to it.
Corrections the verifiers made to the original report:

> The defect is real as described: for MultiBatch sparse small-T launches the mask row is `2*b*full_width + column_begin + token`, but it should be `b*full_width + column_begin + token`. The fix is `mask_base = column_base`, which restores the pre-8fd613f3b indexing.
> 
> Two details to adjust in the reviewer's account:
> 1. A fault is unlikely. The mask is a work_ arena allocation, and gqa_attention then opens a workspace scope and allocates the small-T partial acc/m/l buffers right after it. The out-of-range rows therefore land mostly inside the same device arena, on buffers the kernel writes concurrently. The practical effect is garbage or another lane's block selection (silently wrong attention for lanes 1..B-1), not an illegal-address fault.
> 2. The scope is wider than stated:
>    - it also covers FP8 KV caches (same kernel with CacheT=uint8_t);
>    - it covers verify widths up to 6 (rows off by b*full_width for every b>0);
>    - it covers the uncommitted GLM indexer, which returns a mask with the same layout.
>    It does not cover INT8 KV, whose decode kernel never receives the selection.

_Line numbers quoted by the verifiers may refer to a later tree; the permalinks above are pinned to `7f123876`._

**Comment by flaviusburca** [comment]
Fixed in commit `ff3685c472289524152cfe2b8fce0b253f790fab`. The fix is committed locally and validated; publication to the remote repository is pending.

Batched sparse attention now applies the batch offset only once when locating block-mask rows. Additional FP8 and six-token verification coverage is in commit `38ec60c1ecf864c13e1ccf14e39835090c38e34c`.

Validation: Sparse attention regressions pass for BF16 and FP8, batch sizes 1/2/4, and widths 1/4/6/16. CUDA memcheck reports no errors; the full GQA correctness suite also passes.
